clash-lib/src/Clash/Core/Evaluator/Semantics.hs

```diff
@@ -0,0 +1,581 @@
+when evaluated. When evaluating a term, constructors and primitives will appear
+either on their own or nested in a sequence of App / TyApp nodes.
+
+The special functions, evaluateDataWith and evalautePrimWith handle the first
+case. As these are leaves in the term AST (they have no subterms), they are
+only values if they take no arguments, otherwise they are eta-expanded, and
+the newly created binders are used to provide variables for each argument.
+
+To identify sequences of applications which are applied to constructors or
+primitives, the evaluateAppWith function unwraps all applications. If it is
+a constructor or primitive, it checks that all arguments are applied and
+eta-expands if any are missing.
+
+There is one more consideration which applies to primitives only: the result
+of a primitive may potentially be a function (e.g. primtives that return an
+action in IO). This is handled in evaluateAppWith, by checking if more
+arguments are given than the primitive requires, and partitioning the
+arguments. The arguments for the primitive are applied and the primitive is
+evaluated, then the remaining arguments are applied.
+-}
+
+isFullyApplied :: Type -> [Either Term Type] -> Bool
+isFullyApplied ty args =
+  equalLength args (fst $ splitFunForallTy ty)
+
+etaExpand :: Term -> Eval Term
+etaExpand x =
+  case collectArgs x of
+    y@(Data dc, _) -> do
+      tcm <- getTyConMap
+      expand tcm (dcType dc) y
+
+    y@(Prim p, _) -> do
+      tcm <- getTyConMap
+      expand tcm (primType p) y
+
+    _ -> pure x
+ where
+  expand :: TyConMap -> Type -> (Term, [Either Term Type]) -> Eval Term
+  expand tcm ty (tm, args) = do
+    let missingArgTys = fst $ splitFunForallTy (applyTypeToArgs tm tcm ty args)
+    missingArgs <- traverse etaNameOf missingArgTys
+
+    pure $ mkAbstraction
+      (mkApps x (fmap (bimap Var VarTy) missingArgs))
+      missingArgs
+
+  etaNameOf :: Either TyVar Type -> Eval (Either Id TyVar)
+  etaNameOf = either (pure . Right) (fmap Left . mkUniqueId "eta")
+
+-- | Default implementation for looking up a variable in the environment.
+-- This checks both the local and global environments, and inlines global
+-- binders which are not marked as NOINLINE.
+--
+-- As variables can point to unevalated terms or WHNF values, this function
+-- also forces unevaluated terms to WHNF when performing lookups. This is
+-- currently performed in an environment without the given identifier, to
+-- prevent expressions like
+--
+--   let x = x in ...
+--
+-- looping if x is looked up, as it exists in the environment as
+--
+--   Var x |-> Var x
+--
+-- TODO: In the future this should not delete identifiers from the environment
+-- if they are terminating. This will allow functions like fmap to be fully
+-- unrolled where possible.
+--
+evaluateVarWith :: (Term -> Eval Value) -> Id -> Eval Value
+evaluateVarWith eval i =
+  go (VNeu (NeVar i)) [evalLocal, evalGlobal]
```

Comment:
The `Id` contains a field telling us whether it refers to a global or local variable. So we should predicate on that, instead of trying local first, followed by trying on global.

Reply:
Yes! I kept feeling as though I was missing something obvious that would make this function nicer to implement